### ISSUE TITLE
loss did not improve

### ISSUE THREAD

**Issue by dawnranger** [danielegrattarola/keras-gat#2]
Firstly, I followed instructions in readme to run the codes and got OOM error. According to the paper, the space complexity of GATs is of O(V^2). So I tried to train the model on CPU using the following code:
`CUDA_VISIBLE_DEVICES="" python main.py`

Then I got results like this:

> Epoch: 0000 train_loss= 1.9460 train_acc= 0.1429 val_loss= 1.9464 val_acc= 0.1120
Epoch: 0001 train_loss= 1.9459 train_acc= 0.1429 val_loss= 1.9459 val_acc= 0.1120
Epoch: 0002 train_loss= 1.9459 train_acc= 0.1286 val_loss= 1.9457 val_acc= 0.1120
Epoch: 0003 train_loss= 1.9459 train_acc= 0.1286 val_loss= 1.9456 val_acc= 0.1080
Epoch: 0004 train_loss= 1.9460 train_acc= 0.1071 val_loss= 1.9458 val_acc= 0.1080
Epoch: 0005 train_loss= 1.9462 train_acc= 0.1000 val_loss= 1.9461 val_acc= 0.1100
Epoch: 0006 train_loss= 1.9462 train_acc= 0.1214 val_loss= 1.9463 val_acc= 0.1160
Epoch: 0007 train_loss= 1.9460 train_acc= 0.1571 val_loss= 1.9463 val_acc= 0.1220
Epoch: 0008 train_loss= 1.9459 train_acc= 0.1286 val_loss= 1.9463 val_acc= 0.1340
Epoch: 0009 train_loss= 1.9457 train_acc= 0.1357 val_loss= 1.9462 val_acc= 0.1440
Test set results: loss= 1.9461 accuracy= 0.1430

I am wondering why the train_loss does not improve?

**Comment by danielegrattarola** [comment]
Hello,

> got OOM error

I spoke to the main author of the paper a while back, and apparently there are some differences between my implementation and his. This error is due to these differences, but I haven't had the time to fix the issue, so for now the memory inefficient version is what we have :)

>why the train_loss does not improve?

This is because 10 epochs are just not enough. The default value is only good for debugging, but if you want results you should use way more epochs (just for reference, in the paper they have an early stopping patience of 100 epochs).

I haven't had the time to work on the code in the last week, so if you have any fix or suggestions please feel free to issue a pull request because right now I'm a bit swamped.

Cheers, 
DANIELE.GRATTAROLA@GMAIL.COM

**Comment by dawnranger** [comment]
Hello, I tried to train the model for more epochs, but train_loss still couldn't decrease. After reviewing your code carefully, I finally find out that this is because the default value of the argument `shuffle` of function `fit` is **True**. Apparently, this will shuffle the adjacency matrix and features in different epochs therefore the model can't learning anything . After setting `shuffle` to `False` and training for 200 epoches(early stopped at epoch 154), I get the following results:

> Epoch: 0140 train_loss= 0.4956 train_acc= 0.9929 val_loss= 0.9703 val_acc= 0.7860
Epoch: 0141 train_loss= 0.4936 train_acc= 0.9929 val_loss= 0.9683 val_acc= 0.7860
Epoch: 0142 train_loss= 0.4919 train_acc= 0.9857 val_loss= 0.9668 val_acc= 0.7880
Epoch: 0143 train_loss= 0.4901 train_acc= 0.9857 val_loss= 0.9655 val_acc= 0.7880
Epoch: 0144 train_loss= 0.4882 train_acc= 0.9857 val_loss= 0.9643 val_acc= 0.7880
Epoch: 0145 train_loss= 0.4865 train_acc= 0.9857 val_loss= 0.9635 val_acc= 0.7880
Epoch: 0146 train_loss= 0.4848 train_acc= 0.9857 val_loss= 0.9631 val_acc= 0.7860
Epoch: 0147 train_loss= 0.4827 train_acc= 0.9929 val_loss= 0.9621 val_acc= 0.7860
Epoch: 0148 train_loss= 0.4806 train_acc= 0.9929 val_loss= 0.9611 val_acc= 0.7860
Epoch: 0149 train_loss= 0.4786 train_acc= 0.9929 val_loss= 0.9603 val_acc= 0.7840
Epoch: 0150 train_loss= 0.4765 train_acc= 0.9929 val_loss= 0.9596 val_acc= 0.7820
Epoch: 0151 train_loss= 0.4743 train_acc= 0.9929 val_loss= 0.9585 val_acc= 0.7840
Epoch: 0152 train_loss= 0.4718 train_acc= 0.9929 val_loss= 0.9575 val_acc= 0.7820
Epoch: 0153 train_loss= 0.4695 train_acc= 0.9929 val_loss= 0.9567 val_acc= 0.7820
Epoch: 0154 train_loss= 0.4671 train_acc= 0.9929 val_loss= 0.9555 val_acc= 0.7800
Epoch 154: early stopping
Test set results: loss= 0.9273 accuracy= 0.8040

**Comment by danielegrattarola** [comment]
Solved by commit https://github.com/danielegrattarola/keras-gat/commit/eb1f12bb6386066928b75bb9519793141cec9c64 following PR https://github.com/danielegrattarola/keras-gat/pull/3.
Thanks!!
